pkg/pillar/cmd/volumemgr/handleverifier.go

```diff
@@ -88,63 +99,120 @@ func MaybeRemoveVerifyImageConfig(ctx *volumemgrContext, objType, imageSha strin
+func MaybeDeleteVerifyImageConfig(ctx *volumemgrContext, objType, imageSha string) {
+
+	log.Infof("MaybeRemoveVerifyImageConfig(%s) for %s", imageSha, objType)
+
+	m := lookupVerifyImageConfig(ctx, objType, imageSha)
+	if m == nil {
+		log.Infof("MaybeDeleteVerifyImageConfig: config missing for %s",
+			imageSha)
+		return
+	}
+	if m.Expired {
+		log.Warnf("MaybeDeleteVerifyImageConfig: already Expired for %s",
+			imageSha)
+		return
+	}
+	if m.RefCount != 0 {
+		log.Warnf("MaybeDeleteVerifyImageConfig: Attempting to delete but not zero "+
+			"RefCount %d. Image Details - Name: %s, "+
+			"ImageSha256:%s",
+			m.RefCount, m.Name, m.ImageSha256)
+		return
+	}
+	m.Expired = true
+	log.Infof("MaybeDeleteVerifyImageConfig done for %s", imageSha)
+}
+
 func handleVerifyImageStatusModify(ctxArg interface{}, key string,
 	statusArg interface{}) {
 	status := statusArg.(types.VerifyImageStatus)
 	ctx := ctxArg.(*volumemgrContext)
 	log.Infof("handleVerifyImageStatusModify for ImageSha256: %s, "+
-		" RefCount %d", status.ImageSha256, status.RefCount)
+		" RefCount %d Expired %t", status.ImageSha256, status.RefCount,
+		status.Expired)
+
+	// We handle two special cases in the handshake here
+	// 1. verifier added a status with RefCount=0 based on
+	// an existing file. We echo that with a config with RefCount=0
+	// 2. verifier set Expired in status in response to us setting Expired
+	// in config. If we still have Expired and RefCount zero in the config
+	// we delete the config, which will result in the file being deleted
+	// in the verifier.
+
+	config := lookupVerifyImageConfig(ctx, status.ObjType, status.ImageSha256)
+	if config == nil && status.RefCount == 0 {
+		log.Infof("handleVerifyImageStatusModify adding RefCount=0 config %s",
+			key)
+
+		// Note: signature-related fields are filled in when
+		// RefCount increases from zero in MaybeAddVerifyImageConfig
+		n := types.VerifyImageConfig{
+			Name:         status.Name,
+			ImageSha256:  status.ImageSha256,
+			Size:         status.Size,
+			FileLocation: status.FileLocation,
+			RefCount:     0,
+		}
+		publishVerifyImageConfig(ctx, status.ObjType, &n)
+		return
+	}
+
+	// If we still publish an Expired config with RefCount == 0 we unpublish it.
+	// If config is not Expired it means it was recreated and we
+	// ignore the Expired status
+	if status.Expired && config != nil && config.RefCount == 0 && config.Expired {
+		log.Infof("handleVerifyImageStatusModify delete config for %s",
+			key)
+		unpublishVerifyImageConfig(ctx, status.ObjType, config.Key())
+	} else if status.Expired {
+		log.Infof("handleVerifyImageStatusModify ignore expired VerifyImageStatus; config not Expired for %s",
+			key)
+	}
 	// Ignore if any Pending* flag is set
 	if status.Pending() {
 		log.Infof("handleVerifyImageStatusModify skipped due to Pending* for"+
 			" ImageSha256: %s", status.ImageSha256)
 		return
 	}
-
```

Comment:
Why are we deleting all of this? Do we not want to update the `BlobStatus` when we get an updated `VerifyImageStatus`? If we do not, then when we call `updateStatus()` and from there to `doUpdateContentTree()`, it will have an out-of-date `BlobStatus` which will require cycling through again.

Reply:
The flow is that updateStatus does a lookupFoo and updates its state. The handleFooModify does not do that; just duplication hence a source of bugs. We already got rid of one of those the other day in handleDownloader; didn't realize it was also in handleverifier at the time.

Reply:
Makes sense.